File: guests/ocaml/tooltips.ml
(* The tooltips scene, OCaml port — guests/rust/tooltips.rs,
   tools/scenes/tooltips.steps, docs/tooltip-plan.md. *)

open Kaya_wire
open Kaya_app

type account = { name : string; note : string } [@@deriving kaya_gen]

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
      let name_help = signal (Str "Your full name as it appears on the card") in
      let accounts = collection_of account_record in

      let root =
        column ~help:"The settings for this account" ~a11y_id:"settings"
          [
            button ~text:"Save" ~help:"Saves the draft to disk"    (* button#0 *)
              ~a11y_id:"save"
              ~on_click:(fun () ->
                write name_help (Str "Your name, as saved"));
            button ~text:"Discard" ~help:"Throws the draft away"   (* button#1 *)
              ~a11y_hint:"discard every change" ~a11y_id:"discard";
            entry ~help_bind:name_help ~a11y_id:"fullname";        (* entry#0 *)
            slider ~help:"How loud the preview plays"              (* slider#0 *)
              ~a11y_id:"volume" ~min:0.0 ~max:1.0 ~value:0.5;
            (* THE TRAILING [()] IS LOAD-BEARING inside an `each` body,
               where nothing checks the result: without it the label is a
               partial application the body discards and no copy stamps
               (docs/traps.md, the curried-children entry). *)
            each (record_handle accounts) (fun () ->
                Tpl.(
                  label ~bind_field:account_name ~help_field:account_note
                    ~a11y_id_field:account_name ()));
          ]
          ()
      in
      mount root;

      insert_record accounts (Str "a")
        { name = "a"; note = "The first account, opened in March" };
      insert_record accounts (Str "b")
        { name = "b"; note = "The second account, opened in May" });

  exit (run app)

File: guests/swift/tooltips.swift
// The tooltips scene, Swift port — guests/rust/tooltips.rs,
// tools/scenes/tooltips.steps, docs/tooltip-plan.md.

import Foundation

struct Account: KayaGen {
    var name: String
    var note: String
}

let app = KayaApp()

app.build { tx in
    let nameHelp = tx.signal(.str("Your full name as it appears on the card"))
    let accounts = accountCollection(tx)

    let settings = tx.column {
        let save = tx.button("Save") { tx in  // button#0
            tx.write(nameHelp, .str("Your name, as saved"))
        }
        tx.setHelp(save, "Saves the draft to disk")
        tx.setA11yId(save, "save")
        let discard = tx.button("Discard")  // button#1
        tx.setHelp(discard, "Throws the draft away")
        tx.setA11yHint(discard, "discard every change")
        tx.setA11yId(discard, "discard")
        let name = tx.entry()  // entry#0
        tx.setHelp(name, nameHelp)
        tx.setA11yId(name, "fullname")
        let volume = tx.slider(min: 0.0, max: 1.0, value: 0.5)  // slider#0
        tx.setHelp(volume, "How loud the preview plays")
        tx.setA11yId(volume, "volume")
        for row in accounts.rows {
            let label = row.label(row.name)
            row.t.setHelp(label, row.note)
            row.t.setA11yId(label, row.name)
        }
    }
    tx.setHelp(settings, "The settings for this account")  // column#0
    tx.setA11yId(settings, "settings")
    tx.mount(settings)

    accounts.insert(tx, .str("a"),
                    Account(name: "a", note: "The first account, opened in March"))
    accounts.insert(tx, .str("b"),
                    Account(name: "b", note: "The second account, opened in May"))
}

app.run()
